0016-base-primitive-unification.md

```diff
@@ -0,0 +1,115 @@
+| **RFC #**         | 0016                                         |
+| **Authors**       | Lev Bishop (lsbishop@us.ibm.com)             |
+|                   | Ian Hincks (ian.hincks@ibm.com)              |
+|                   | Blake Johnson (blake.johnson@ibm.com)        |
+|                   | Chris Wood (cjwood@us.ibm.com)               |
+| **Submitted**     | 2023-10-26                                   |
+| **Updated**       | YYYY-MM-DD     
+
+## Summary
+
+The primitives form an execution framework for QPUs and their simulators.
+Each primitive specializes at some computational task.
+For example, the `Estimator` specializes at computing expectation values of user-provider observables.
+
+Different types of primitives are implemented as different abstract subclasses of `BasePrimitive` (e.g. `BaseEstimator`), and different primitive implementations as further subclasses (e.g. `qiskit_aer.primitives.Estimator`).
+Because the constructor of these subclasses is reserved for implementation-specific setup, the main entry-point to execution is the `.run()` method of a particular class.
+However, as it stands, `BasePrimitive` does not provide `.run()` as an abstraction; each type of primitive gets to define its own signature.
+
+This RFC proposes an abstraction for `BasePrimitive.run()` as well as structures for its inputs and outputs.
+
+## Motivation
+
+The central motivation of this RFC is to more clearly emphasize that primitives are a well-defined execution framework on top of which applications can be built.
+We aim to force each category of primitive to clearly state what units of quantum work (tasks) it is able to perform, and what outputs can be expected.
+It is understood that having the proposed abstract base method alone will not necessarily improve the day-to-day life of a typical user of the primitives, and neither will it necessarily improve the ease of implementing new primitive types or implementations: this change is mainly about clarifying the shared nature of the primitives, and reflecting this view in abstractions.
+
+## User Benefit
+
+One class of benefited users are application frameworks that are built on top of the primitives, which, following this proposal, will be able to exploit the promise of a homogenous experience across all primitives.
+As an example, `qiskit_experiments` defines a suite of common diagnostic, calibration, and characterization experiments.
+Currently, it is designed around the `Backend.run` execution interface.
+One could imagine that, for example, instead of each `Experiment` providing a `circuits` attribute specifying the circuits to run through `Backend.run`, they could rather each specify a `tasks` attribute indicating which primitive units of work need performed.
+The `qiskit_experiments` execution and analysis machinery could then rely on the abstraction of `BasePrimitive.run` when reasoning about dispatching and collecting results.
```

Comment:
I'm not sure I follow the abstract classes defined below are so broadly defined I can't imagine anything ever saying anything accepted `BasePrimitive` as a type Specifically the estimator and sampler behave quite differently both in the allowed inputs and the data returned. Like for `qiskit-experiments` do you have a specific case where you think this applies?